source/s3_client.c

```diff
@@ -930,253 +909,179 @@ static void s_s3_client_process_work_task(struct aws_task *task, void *arg, enum
-            s_s3_vip_connection_destroy(client, vip_connection);
-            aws_linked_list_push_back(&work_destroy_list, &work->node);
-            continue;
-        }
+    uint32_t num_meta_requests = 0;
 
-        struct aws_linked_list_node *next_meta_request_node = NULL;
+    for (struct aws_linked_list_node *node = aws_linked_list_begin(&client->threaded_data.meta_requests);
+         node != aws_linked_list_end(&client->threaded_data.meta_requests);
+         node = aws_linked_list_next(node)) {
+        ++num_meta_requests;
+    }
 
-        /* If this VIP connection has a meta request, clear it out, and try to grab the next one in the list. */
-        if (vip_connection->threaded_data.meta_request != NULL) {
-            next_meta_request_node =
-                aws_linked_list_next(&vip_connection->threaded_data.meta_request->threaded_data.node);
+    while (!aws_linked_list_empty(&vip_connections_updates)) {
 
-            aws_linked_list_remove(&vip_connection->threaded_data.meta_request_reference_node);
-            aws_s3_meta_request_release(vip_connection->threaded_data.meta_request);
-            vip_connection->threaded_data.meta_request = NULL;
-        }
+        struct aws_linked_list_node *node = aws_linked_list_pop_front(&vip_connections_updates);
+        struct aws_s3_vip_connection *vip_connection = AWS_CONTAINER_OF(node, struct aws_s3_vip_connection, node);
 
-        bool has_next_meta_request_node =
-            next_meta_request_node != NULL &&
-            next_meta_request_node != aws_linked_list_end(&client->threaded_data.meta_requests);
+        size_t cleaning_up = aws_atomic_load_int(&vip_connection->owning_vip->cleaning_up);
 
-        /* If we don't have a next-meta-request node (either we didn't have a meta request, or it's next-node didn't get
-         * us anything), and we do have something in our meta request list, then start at the beginning of the list.*/
-        if (!has_next_meta_request_node && !aws_linked_list_empty(&client->threaded_data.meta_requests)) {
-            next_meta_request_node = aws_linked_list_begin(&client->threaded_data.meta_requests);
+        /* If this VIP connection is pending destruction, go ahead and clean it up now. */
+        if (cleaning_up) {
+            s_s3_vip_connection_destroy(client, vip_connection);
+            continue;
         }
 
-        /* If we still don't have a meta request node, there's nothing for this VIP connection to do. */
-        if (next_meta_request_node == NULL) {
-            AWS_ASSERT(vip_connection->threaded_data.meta_request == NULL);
-            aws_linked_list_push_back(&client->threaded_data.idle_vip_connections, &work->node);
+        AWS_ASSERT(vip_connection->request == NULL);
+
+        /* If we don't have any meta requests or how many requests we have in flight is over the maximum allowed, put
+         * the vip connection into the idle list. */
+        if (aws_linked_list_empty(&client->threaded_data.meta_requests) ||
```

Comment:
How big is this list? Could we use an array_list of pointers instead? Looping over this list twice in this function to count it is intentionally inducing likely cache misses just to produce a count. Might create a death by paper cut situation with enough requests in flight.

Reply:
Woops, the loop that always goes through the entire list was just for logging--I need to take that out.  The list is for all of the meta requests being processed.

This part is just looking for the next meta request that has requests, which we only do for the number of vip connections that need work.  It is possible that all of the meta requests in the list ran out of work at once, in which case we would go through the entire list.  We could change it over to an arraylist.

Reply:
If that's all, don't sweat it, I just didn't want this to be on the hot path.